### ISSUE TITLE
Add possibility to process data from a datasource before generation starts.

### ISSUE THREAD

**Issue by MSDelorme** [eclipse-birt/birt#1351]
There are quite some use-cases where you want to be able to determine certain properties of the output based on information from a data source. A number of things are processed before a connection with data-source is established.
Some examples:
- output file name
- output type
- master page background image

The only possibility to get data from a data source earlier in the process, is using a dynamic parameter.  However when the runtime engine is used this is not a viable way. 

Would it be possible to have some init process step where you can connect a data-source to determine these kind of things?

**Comment by MSDelorme** [comment]
In addition, for some things you can make a working around (not the once mention in my first message). Adding a grid , connect to a data source and hide this element. This is however not a very elegant solution IMHO.
In general I would say that is very useful to able able to determine variables based on a data source without having to add items to a a designs body and before generation is started.

**Comment by hvbtup** [comment]
I don't have a solution, but I have to admit this is an interesting question. I was in the same situation several times.

I would say that for *most* things you can use the workaround you mentioned (with a hidden layout element bound to a data set), and I think that this workaround is pretty reasonable. I tend to use red font color to mark these "logic in the layout" and our reports usually don't output red text, so the developer sees immediately that some logic is happening here.

This workaround doesn't work in all cases.
E.g. once I wanted to modify and populate a grid (including removing columns etc), with input coming from the DB.
I was able to use the DE API to basically clone the existing Data Source and create a Data Set in the beforeFactory(?) event of the report, but while I was able to create a connection, I was not able to read the report parameter values.
I remember that I thought: If only there was a way to open and execute a data set from Javascript within the report (at a stage where I also can access the report parameters).

So I'm +1 for adding such an API. But I'm not able to provide it myself.

**Comment by hyrahul** [comment]
Hi @hvbtup and @wimjongman , I'd like to work on this issue. the hidden gird work around makes sense for simple cases, I'll dig into the engine's pre-generation lifecycle and come back.

**Comment by SteveSchafer-Innovent** [comment]
+1 on this.  I have many reports with hidden tables that exist only to gather some data.

**Comment by hvbtup** [comment]
Just wanted to say that changing anything in the execution sequence of a report has a high risk of breaking things. If you choose to work on this, please make sure that the default behavior of reports is not changed.

What @SteveSchafer-Innovent wants can also be achieved with a list, or, in some cases, even with a single dynamic text item bound to the data set. A list is better suited than a table, because it has less overhead.
One thing that is annoying with lists, though, is that they always reserve screen space in the BIRT designer for the header and the footer area, even if these areas are empty.
It would be nice if we could just optionally avoid displaying these empty areas. This would only effect the representation in the designer. That would be worth a separate issue.

**Comment by hyrahul** [comment]
Dug into the engine most of this is already possible today, just undocumented:
- `IReportEngine.createDatasetPreviewTask()` already lets you run a data set query with report parameters resolved **before** calling the real run/render task.
- That covers:
   - output file name
   - output type
   - master page background image
   - all from the calling application, with **zero changes** to the report generation sequence.
where it falls short: when you don't control the calling application, only the `.rptdesign` itself (e.g. modifying a grid's column from DB data, inside `beforeFactory`).

For that, proposing one small addition:
- A script-callable `reportContext.openDataSet(name)`, usable from `beforeFactory`
- Reuses the report's own already-open data connection and already-resolved parameters
- No second connection, no execution-order change

On the points raised above:
- **Execution sequence risk** - agreed, and both parts of this proposal are additive only: nothing about when/how `beforeFactory` or the run/render lifecycle fires changes or anyone not explicitly calling the new API.
- **List / single dynamic text item instead of a hidden table** - good point, light-weight that a table for the same workaround. Doesn't remove the need for a workaround though, which is what this proposal trying to avoid entirely.

Let me know if this sounds reasonable, or if I'm off base somewhere.

**Comment by hvbtup** [comment]
What would `reportContext.openDataSet (name)` do, how would it be used?

OT: Personally, I prefer human-written text. Your texts look like mostly AI generated.

**Comment by hyrahul** [comment]
> What would `reportContext.openDataSet (name)` do, how would it be used? -> reportContext.openDataSet(Mydataset) it will return a result set which we can iterate in script and help us to build the all the variables required before the layout got built.

> 
> OT: Personally, I prefer human-written text. Your texts look like mostly AI generated -> i use the AI to format my message before posting in chat.

**Comment by wimjongman** [comment]
@hyrahul thanks for the input.

Can you supply or describe an example script on that exit point so we have an idea what you are thinking of?

**Comment by hyrahul** [comment]
> [@hyrahul](https://github.com/hyrahul) thanks for the input.
> 
> Can you supply or describe an example script on that exit point so we have an idea what you are thinking of?

script can be e.g. before layout exist we can decide report parameter or structure 
var rs = reportContext.openDataSet("RegionConfig");
var regions = [];
while(rs.next()){
regions.push(rs.getString("REGION_NAME"));
}
rs.close();

reportContext.setPersitentGlobalValriable("regionCount", regions.length);
reportContext.setPersitentGlobalValriable("regions", region.);

layout will uses these variables in visibility expression on options sections e.g. columns, headers, a title.

hidden table runs during factory, too late to add anything which is already evaluated earlier.

**Comment by hvbtup** [comment]
I'm not sure. What _I_ would need to make this really helpful, would be the ability to feed the known report parameter values into the data sets and then to use the data set results to modify the design.

I think what I need _requires_ a change in the processing sequence like this:
1) Read and parse the XML.
2) With access to the data sets and other report parameters, get and modify report parameters.
3) Modify the report items of body and master pages, using the data sets and report parameters
4) Execute the body and master pages.

All of this works, but not in this order.
I can use the DE API for 3), but only in the onPrepare event, but at this point in time we don't have access to the report parameters and/or data sets yet. And in beforeFactory, I can't modify the design anymore.

**Comment by hyrahul** [comment]
> I'm not sure. What _I_ would need to make this really helpful, would be the ability to feed the known report parameter values into the data sets and then to use the data set results to modify the design.
> 
> I think what I need _requires_ a change in the processing sequence like this:
> 
> 1. Read and parse the XML.
> 2. With access to the data sets and other report parameters, get and modify report parameters.
> 3. Modify the report items of body and master pages, using the data sets and report parameters
> 4. Execute the body and master pages.
> 
> All of this works, but not in this order. I can use the DE API for 3), but only in the onPrepare event, but at this point in time we don't have access to the report parameters and/or data sets yet. And in beforeFactory, I can't modify the design anymore.

@hvbtup 
Traced this further parameters are resolved before **onPrepare** runs usingParameterValue() and it happens before **prepareDesign()** in the task sequence by the time onPrepare get fires parameter values are already set. tested this with default value and with a parameter it resolved as expected inside onPrepare.

I added a small openDataSet(regionConfig) helper to test with and called it from an onPrepare script it read real rows from the db same as it would from **beforeFactory**

so both pieces as you mentioned (report parameters, data set results) already be available at onPrepare. which is also where DE API modification work as well.

Did you run into something specific that didn't work there ?

[onPrepare] read 27 rows -> it is from dataset query against the sample db (RegionConfig) 
[onPrepare] param = hello-from-test -> This is a scalar parameter with a default value read from same `onPrepare` script.

```
<parameters>
        <scalar-parameter name="testParam" id="12000">
            <property name="dataType">string</property>
            <property name="paramType">simple</property>
            <property name="isRequired">false</property>
        </scalar-parameter>
    </parameters>

    <method name="onPrepare"><![CDATA[
java.lang.System.out.println("[onPrepare] starting");
try {
    var rs = reportContext.openDataSet("RegionConfig");
    var n = 0;
    while (rs.next()) { n++; }
    rs.close();
    java.lang.System.out.println("[onPrepare] read " + n + " rows");
} catch (e) {
    java.lang.System.out.println("[onPrepare] FAILED: " + e);
}
java.lang.System.out.println("[onPrepare] param = " + reportContext.getParameterValue("testParam"));
]]></method>
```

**Comment by hvbtup** [comment]
In fact you're right: It is already possible to access parameter values and modify the layout in the reports' `beforeFactory` and `onPrepare` events. I tested this with the current master branch and with BIRT 4.2.1 (yes!).

What's missing is just the ability to also use result sets from data sets, which is what you propose. 

I remember I wanted to create a grid based on some DB data depending on the report parameters. This didn't work, and the reason obviously was that something like `openDataSet` is not available. 
So my memory faked me into believing that I couldn't access the DB at all early enough to be allow modifying the layout.
But probably I could have accessed _a single value_ from the DB, by using a cascading parameter group or sth like that, and get the default value for P2 depending on P1, then using this P2 default value for modifying the layout.

Well, I'm looking forward to your PR.

**Comment by hvbtup** [comment]
> But probably I could have accessed _a single value_ from the DB, by using a cascading parameter group or sth like that, and get the default value for P2 depending on P1, then using this P2 default value for modifying the layout.

Checked this: This does not work. So, we would need something like openDataSet even for accessing only a single value from the DB based on a report parameter.

**Comment by hyrahul** [comment]
Draft PR for #1351
### Summary
- Added `reportContext.openDataSet(?)` so `beforeFactory` / `onPrepare` scripts can read data-set rows using the report's own data engine and resolved parameters.
- Verified the lifecycle with the sample database: parameters are resolved before `onPrepare`, and data sets are available at that point.
- Added tests covering:
  - Reading a data set and using its value to change a label.
  - A non-existent data set.
  - A script that doesn't close the result.
  - A parameterized data set, confirming the resolved parameter value reaches the query correctly.
- Also moved `needOutputResultSet` earlier, since it was previously set after `startFactory` and could be missed when the data engine was opened from a script.

Keeping this as a draft for now since this introduces a new API. I'd appreciate feedback on the API shape before finalizing.

**Comment by wimjongman** [comment]
Can this be closed?
